server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -422,16 +424,17 @@ public void onFailure(Exception e) {
             cancellableThreads.checkForCancel();
             ensureNotRelocatedPrimary(primary);
             action.accept(l2);
-        })), ThreadPool.Names.GENERIC);
+        })), threadPool.executor(ThreadPool.Names.GENERIC));
     }
 
     static void runUnderPrimaryPermit(
+        ThreadPool threadPool,
```

Comment:
Ooh interesting I wasn't expecting this. I would prefer instead to fix up `IndexShardOperationPermits` to work on an `Executor` instead of its name, in a PR that we merge before this one?

Reply:
Yeah I think we should chase this change all the way up the call tree too - see https://github.com/elastic/elasticsearch/pull/99719. If we do that first then we shouldn't need to add a `ThreadPool` parameter here.